crypto/dh/dh_key.c

```diff
@@ -86,26 +86,50 @@ static int compute_key(unsigned char *key, const BIGNUM *pub_key, DH *dh)
         goto err;
     }
 
-    ret = BN_bn2bin(tmp, key);
+    /* return the padded key, i.e. same number of bytes as the modulus */
+    ret = BN_bn2binpad(tmp, key, BN_num_bytes(dh->params.p));
  err:
     BN_CTX_end(ctx);
     BN_CTX_free(ctx);
     return ret;
 }
 
+/*-
+ * NB: This function is inherently not constant time due to the
+ * RFC 5246 (8.1.2) padding style that strips leading zero bytes.
+ */
 int DH_compute_key(unsigned char *key, const BIGNUM *pub_key, DH *dh)
 {
+    int ret = 0, i;
+    volatile size_t npad = 0, mask = 1;
+
+    /* compute the key; ret is constant unless compute_key is external */
 #ifdef FIPS_MODULE
-    return compute_key(key, pub_key, dh);
+    ret = compute_key(key, pub_key, dh);
 #else
-    return dh->meth->compute_key(key, pub_key, dh);
+    ret = dh->meth->compute_key(key, pub_key, dh);
 #endif
+    if (ret <= 0)
+        return ret;
+
+    /* count leading zero bytes */
+    for (i = 0; i < ret; i++) {
+        mask &= !key[i];
+        npad += mask;
+    }
+
+    /* unpad key */
+    ret -= npad;
+    memmove(key, key + npad, ret);
```

Comment:
Could you please also add the zeroization of the tail bytes after moving the pad?

Reply:
I left it out because it seemed irresponsible to operate on a secret in a non-CT way when you don't have to.

For my understanding, is there some reason I'm missing?

Reply:
is it because you think `ret + npad` writes will hide better than `ret` bytes?

Reply:
No, my concern is leaving some bytes of the computed key on the heap without proper zeroization. The caller is expected to zeroize the key when not used anymore however he does not know that there were a few bytes more at the end after the pad removal.

Reply:
Fair enough. Fixed in 0a50ca30a08a943984e9c34d2357ecfbe59db8ce